tests/stats.c

```diff
@@ -0,0 +1,95 @@
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+    IN THE SOFTWARE.
+*/
+
+#include "../src/nn.h"
+#include "../src/reqrep.h"
+
+#include "testutil.h"
+
+#define SOCKET_ADDRESS "tcp://127.0.0.1:5555"
+
+int main (int argc, char **argv)
+{
+    int rc;
+    int rep1;
+    int req1;
+
+    /*  Test req/rep with full socket types. */
+    rep1 = test_socket (AF_SP, NN_REP);
+    test_bind (rep1, SOCKET_ADDRESS);
+    req1 = test_socket (AF_SP, NN_REQ);
+    test_connect (req1, SOCKET_ADDRESS);
+    nn_sleep (200);
+
+    nn_assert (nn_get_statistic(rep1, NN_STAT_ACCEPTED_CONNECTIONS) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_ESTABLISHED_CONNECTIONS) == 0);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_CURRENT_CONNECTIONS) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_SENT) == 0);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_RECEIVED) == 0);
+
+    nn_assert (nn_get_statistic(req1, NN_STAT_ACCEPTED_CONNECTIONS) == 0);
+    nn_assert (nn_get_statistic(req1, NN_STAT_ESTABLISHED_CONNECTIONS) == 1);
+    nn_assert (nn_get_statistic(req1, NN_STAT_CURRENT_CONNECTIONS) == 1);
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_SENT) == 0);
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_RECEIVED) == 0);
+
+    test_send (req1, "ABC");
+    nn_sleep (100);
+
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_SENT) == 1);
+    nn_assert (nn_get_statistic(req1, NN_STAT_BYTES_SENT) == 3);
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_RECEIVED) == 0);
+    nn_assert (nn_get_statistic(req1, NN_STAT_BYTES_RECEIVED) == 0);
+
+    test_recv(rep1, "ABC");
+
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_SENT) == 0);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_BYTES_SENT) == 0);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_RECEIVED) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_BYTES_RECEIVED) == 3);
+
+    test_send (rep1, "OK");
+    test_recv (req1, "OK");
+
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_SENT) == 1);
+    nn_assert (nn_get_statistic(req1, NN_STAT_BYTES_SENT) == 3);
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_RECEIVED) == 1);
+    nn_assert (nn_get_statistic(req1, NN_STAT_BYTES_RECEIVED) == 2);
+
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_SENT) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_BYTES_SENT) == 2);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_RECEIVED) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_BYTES_RECEIVED) == 3);
+
+    test_close (req1);
+
+    nn_sleep (100);
+
+    nn_assert (nn_get_statistic(rep1, NN_STAT_ACCEPTED_CONNECTIONS) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_ESTABLISHED_CONNECTIONS) == 0);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_CURRENT_CONNECTIONS) == 0);
+
+    test_close (rep1);
+
+    return 0;
+}
+
```

Comment:
Valid point.  I think we should advise caution.  While the API itself can be considered stable, some of the statistics may be considered unstable, and relying upon their values programmatically for purposes other than visibility or debugging is at least strongly discouraged.

That said, this is a level of visibility that can be *useful*.  Imagine knowing that you could monitor connections statistic to determine when to start flooding data... etc.  I do imagine that I'd wind up using this information for the express purpose of test cases, since now we have visibility into some state that we previously lacked.  But I wouldn't want to do any of that in production.